packages/axelar-wasm-std-derive/src/lib.rs

```diff
@@ -21,80 +25,155 @@ pub fn into_contract_error_derive(input: TokenStream) -> TokenStream {
+/// # enum SomeEventWithUnnamedField {
+/// #     Uint,
+/// #     Named { some_uint: u64 },
+/// #     Unnamed(u64),
+/// # }
+/// ```
+#[proc_macro_derive(IntoEvent)]
+pub fn into_event(input: TokenStream) -> TokenStream {
+    let input = syn::parse_macro_input!(input as syn::ItemEnum);
+    let event_enum = input.ident.clone();
+
+    let variant_matches: Result<Vec<_>, _> = input
+        .variants
+        .into_iter()
+        .map(|variant| match &variant.fields {
+            syn::Fields::Named(fields) => Ok(match_named_field(&event_enum, &variant, fields)),
+            syn::Fields::Unit => Ok(match_uint_field(&event_enum, &variant)),
+            syn::Fields::Unnamed(_) => Err(syn::Error::new(
+                Span::call_site(),
+                "unnamed fields are not supported",
+            )),
+        })
+        .collect();
+    let variant_matches = match variant_matches {
+        Ok(variant_matches) => variant_matches,
+        Err(err) => return err.to_compile_error().into(),
+    };
+
+    quote! {
+        impl From<&#event_enum> for cosmwasm_std::Event {
+            fn from(event: &#event_enum) -> Self {
+                match event {
+                    #(#variant_matches),*
+                }
             }
         }
 
-        impl From<#event_struct> for cosmwasm_std::Event {
-            fn from(event: #event_struct) -> Self {
+        impl From<#event_enum> for cosmwasm_std::Event {
+            fn from(event: #event_enum) -> Self {
                 (&event).into()
             }
         }
-    })
+    }
+    .into()
+}
+
+fn match_named_field(event_enum: &Ident, variant: &Variant, fields: &FieldsNamed) -> TokenStream2 {
+    let variant_name = variant.ident.clone();
+    let event_name = variant_name.to_string().to_snake_case();
+    let field_names = fields
+        .named
+        .iter()
+        .map(|field| field.ident.clone().expect("field must have a name"));
+
+    let new_event = quote! {
+        #event_enum::#variant_name { #(#field_names), * } => cosmwasm_std::Event::new(#event_name)
+    };
+    let add_attributes = fields.named.iter().map(|field| {
+        let field_name = field.ident.clone().expect("field must have a name");
+        let field_name_str = field_name.to_string();
+        let attribute_name = field_name_str.to_snake_case();
+
+        quote! {
+            add_attribute(#attribute_name, serde_json::to_string(#field_name).expect(format!("failed to serialize event field {}", #field_name_str).as_str()))
+        }
+    });
+
+    let results = iter::once(new_event).chain(add_attributes);
+
+    quote! {
+        #(#results).*
+    }
+}
+
+fn match_uint_field(event_enum: &Ident, variant: &Variant) -> TokenStream2 {
+    let variant_name = variant.ident.clone();
+    let event_name = variant_name.to_string().to_snake_case();
+
+    quote! {
+        #event_enum::#variant_name => cosmwasm_std::Event::new(#event_name)
+    }
 }
```

Comment:
slightly clearer name and unnecessary clone removed
```suggestion
fn match_unit_variant(event_enum: &Ident, variant_name: &Ident) -> TokenStream2 {
    let event_name = variant_name.to_string().to_snake_case();

    quote! {
        #event_enum::#variant_name => cosmwasm_std::Event::new(#event_name)
    }
}
```